core/src/test/java/org/apache/calcite/test/HepPlannerTest.java

```diff
@@ -242,6 +247,53 @@ private void assertIncludesExactlyOnce(String message, String digest, String sub
     assertThat(listener.getApplyTimes() == 1, is(true));
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-2454">[CALCITE-2454]
+   * HepPlanner should not pick cached RelNode when type mismatch while digest matches</a>. */
+  @Test public void testRelNodeCacheWithDigest() {
+    HepProgramBuilder programBuilder = HepProgram.builder();
+    programBuilder.addRuleInstance(ReduceExpressionsRule.PROJECT_INSTANCE); // dummy rules.
+    programBuilder.addRuleInstance(ReduceExpressionsRule.CALC_INSTANCE);
+    programBuilder.addRuleInstance(ProjectMergeRule.INSTANCE);
+    HepPlanner planner = new HepPlanner(programBuilder.build());
+
+    RelBuilder relBuilder = RelBuilder.create(RelBuilderTest.config().build());
+    RelDataTypeFactory typeFactory = relBuilder.getTypeFactory();
+    RelDataType longType = typeFactory.createSqlType(SqlTypeName.BIGINT);
+    RelDataType intType = typeFactory.createSqlType(SqlTypeName.INTEGER);
+    RelDataType doubleType = typeFactory.createSqlType(SqlTypeName.DOUBLE);
+
+    RelNode values = relBuilder
+        .values(new String[] {"f1", "f2", "f3"}, 1, 2, 3)
+        .build();
+
+    // schema: 1, 2, 3.0D | int, bigint, double
+    RelNode p1 = relBuilder
+        .push(values)
+        .project(relBuilder.getRexBuilder().makeLiteral(1, intType, false),
+            relBuilder.getRexBuilder().makeLiteral(2, longType, false),
+            relBuilder.getRexBuilder().makeLiteral(3.0D, doubleType, false))
+        .build();
+    // schema: 1, 2, 3.0D | bigint, int, double
+    RelNode p2 = relBuilder
+        .push(values)
+        .project(relBuilder.getRexBuilder().makeLiteral(1, longType, false),
+            relBuilder.getRexBuilder().makeLiteral(2, intType, false),
+            relBuilder.getRexBuilder().makeLiteral(3.0D, doubleType, false))
+        .build();
+    RelNode r = relBuilder
+        .push(p1)
+        .push(p2)
+        .union(true)
+        .build();
+    planner.setRoot(r);
+    RelNode bestRel = planner.findBestExp();
+
+    assertThat(r.getInput(0).getRowType().equals(r.getInput(1).getRowType()), is(false));
```

Comment:
@danny0405 , thanks for the test. Would you please add relevant messages so the assertion failure is human readable?

For instance, current test fails (without the fix) as follows:

```
java.lang.AssertionError: 
Expected: is <false>
     but: was <true>
Expected :is <false>
Actual   :<true>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.apache.calcite.test.HepPlannerTest.testRelNodeCacheWithDigest(HepPlannerTest.java:100)
```

It is not really helpful, and it will be really hard to maintain that kind of tests.
Good failure should look like a good bug report.

Reply:
@danny0405 I have added some comment and assert msg, thx for the review.

Reply:
The one that does not require full understanding of the test code in order to understand the failure.

For instance:

```
java.lang.AssertionError: COALESCE(+(CAST(-626799667):INTEGER, null), ?0.notNullInt1, -(null))
coalesce(plus(rexBuilder.makeCall(SqlStdOperatorTable.CAST, literal(-626799667)), nullInt), vIntNotNull(1), unaryMinus(nullInt)) had non-nullable type INTEGER, and it was optimized to null
nullInt that has nullable type INTEGER, unknownAsFalse

	at org.junit.Assert.fail(Assert.java:88)
	at org.apache.calcite.test.fuzzer.RexProgramFuzzyTest.checkUnknownAs(RexProgramFuzzyTest.java:266)
	at org.apache.calcite.test.fuzzer.RexProgramFuzzyTest.checkUnknownAs(RexProgramFuzzyTest.java:163)
	at org.apache.calcite.test.fuzzer.RexProgramFuzzyTest.generateRexAndCheckTrueFalse(RexProgramFuzzyTest.java:412)
	at org.apache.calcite.test.fuzzer.RexProgramFuzzyTest.runRexFuzzer(RexProgramFuzzyTest.java:344)
	at org.apache.calcite.test.fuzzer.RexProgramFuzzyTest.testFuzzy(RexProgramFuzzyTest.java:315)
```

One doesn't even need to look into the test code (and/or understand it) in order to understand the nature of the failure.

Your message says "The two inputs should have different row types after planner promotion", however it is not clear what is promotion, why the types should be different, etc, etc. The actual and expected values are lost, so the only way to understand the test failure is to set a breakpoint in debugger.

Reply:
@vlsi 

> Hep should not change node type after we found best plan for passed in tree nodes, for this case, the bestRel's two projects should keep row type as they originally was, cause data type is an important factor for RelNode equivalence.

What do you thing about this msg ?

Reply:
> What do you thing about this msg ?

@danny0405 , the message is better, however it gives very little information when combined with expected/actual. You lose both expected and actual types, so this long message very little info on the nature of the failure.


Here's what I mean: https://github.com/apache/calcite/pull/1002/commits/d49690660227f99545e3f6bc25afa7f97b43fcc7